### PR TITLE
Added endpoint to get an explain or explain analyze for a query.

### PR DESCRIPTION
**Pull request by PiSoup**

### CODE REVIEW

File: soql-server-pg/src/main/scala/com/socrata/pg/server/QueryServer.scala [hunk 1]
```diff
@@ -278,12 +313,35 @@ class QueryServer(val dsInfo: DSInfo, val caseSensitivity: CaseSensitivity) exte
...
+                  JsonUtil.writeJson(writer, Array(explainInfo))
+                  writer.flush()
+                  writer.close()
               }
           }
       }
     }
   }
 
+  trait QueryResultBall
```
Comment (bahrieinn) [inline soql-server-pg/src/main/scala/com/socrata/pg/server/QueryServer.scala]: 🎱?

File: soql-server-pg/src/main/scala/com/socrata/pg/server/QueryServer.scala [hunk 2]
```diff
@@ -376,8 +453,12 @@ class QueryServer(val dsInfo: DSInfo, val caseSensitivity: CaseSensitivity) exte
             NotModified(Seq(etag))
           case Some(_) | None =>
             try {
-              val (qrySchema, version, results) = runQuery(pgu, copy, analysis, rowCount, queryTimeout)
-              Success(qrySchema, copy.copyNumber, version, results, etag, lastModified)
+              runQuery(pgu, copy, analysis, rowCount, queryTimeout, explain, analyze) match {
+                case RowsQueryResult(qrySchema, version, results) =>
+                  Success (qrySchema, copy.copyNumber, version, results, etag, lastModified)
```
Comment (bahrieinn) [inline soql-server-pg/src/main/scala/com/socrata/pg/server/QueryServer.scala]: I think the whitespace is unintended here?

File: soql-server-pg/src/main/scala/com/socrata/pg/query/DataSqlizerQuerier.scala [hunk 3]
```diff
@@ -133,6 +134,57 @@ trait DataSqlizerQuerier[CT, CV] extends AbstractRepBasedDataSqlizer[CT, CV] wit
...
+    ExplainInfo(
+      sql.toString(),
+      result.mkString
+    )
+  }
+
+  private def explainSql(conn: Connection, pSql: ParametricSql, timeout: Option[Duration], analyze: Boolean): ResultSet = {
+    try {
+      if (timeout.isDefined && timeout.get.isFinite()) {
+        val ms = timeout.get.toMillis.min(Int.MaxValue).max(1).toInt.toString
```
[the remark this answers is not in the document]
Comment (PiSoup) [reply soql-server-pg/src/main/scala/com/socrata/pg/query/DataSqlizerQuerier.scala]: Just clamping the value to 1 < timeout < Int.maxvalue

File: soql-server-pg/src/main/scala/com/socrata/pg/query/DataSqlizerQuerier.scala [hunk 4]
```diff
@@ -133,6 +134,57 @@ trait DataSqlizerQuerier[CT, CV] extends AbstractRepBasedDataSqlizer[CT, CV] wit
...
+        execute(conn, setTimeout(ms))
+      }
+
+      val query = "EXPLAIN " + (if(analyze) "ANALYZE " else "") + pSql.sql.head
+
+      // Statement to be closed by caller
+      val stmt = conn.prepareStatement(query)
+
+      // Explain should never return a massive amount
+      stmt.setFetchSize(0)
```
[the remark this answers is not in the document]
Comment (PiSoup) [reply soql-server-pg/src/main/scala/com/socrata/pg/query/DataSqlizerQuerier.scala]: In that case no

File: soql-server-pg/src/main/scala/com/socrata/pg/server/QueryServer.scala (hunk 1 above)
Comment (PiSoup) [reply soql-server-pg/src/main/scala/com/socrata/pg/server/QueryServer.scala]: I'm bad at names, if you've got a better name for it I'll gladly take it.

File: soql-server-pg/src/main/scala/com/socrata/pg/server/QueryServer.scala (hunk 2 above)
Comment (PiSoup) [reply soql-server-pg/src/main/scala/com/socrata/pg/server/QueryServer.scala]: The newly added match adds a set of curly braces.

**Pull request merged**